src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,374 @@
+    ssize_t _max_bytes;
+    size_t _num_batches{};
+    model::timeout_clock::time_point _deadline;
+    inclusive_offset_range _range;
+    ss::abort_source _as;
+    retry_chain_node _rtc;
+    retry_chain_logger _ctxlog;
+};
+
+ss::input_stream<char> make_reader_input_stream(
+  model::ntp ntp,
+  model::record_batch_reader r,
+  size_t read_buffer_size,
+  inclusive_offset_range range,
+  model::timeout_clock::duration timeout) {
+    auto deadline = model::timeout_clock::now() + timeout;
+    auto ds = std::make_unique<reader_ds>(
+      ntp, std::move(r), read_buffer_size, range, deadline);
+    ss::input_stream<char> s(ss::data_source(std::move(ds)));
+    return s;
+}
+
+segment_upload::segment_upload(
+  ss::lw_shared_ptr<cluster::partition> part,
+  size_t read_buffer_size,
+  ss::scheduling_group sg)
+  : _ntp(part->get_ntp_config().ntp())
+  , _part(part)
+  , _rd_buffer_size(read_buffer_size)
+  , _sg(sg)
+  , _rtc(_as)
+  , _ctxlog(archival_log, _rtc, _ntp.path()) {}
+
+void segment_upload::throw_if_not_initialized(std::string_view caller) const {
+    if (_stream.has_value() == false) {
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format,
+          "Segment upload is not initialized {} to {}",
+          _ntp,
+          caller));
+    }
+}
+
+ss::future<result<std::unique_ptr<segment_upload>>>
+segment_upload::make_segment_upload(
+  ss::lw_shared_ptr<cluster::partition> part,
+  inclusive_offset_range range,
+  size_t read_buffer_size,
+  ss::scheduling_group sg,
+  model::timeout_clock::time_point deadline) {
+    std::unique_ptr<segment_upload> upl(
+      new segment_upload(part, read_buffer_size, sg));
+
+    auto res = co_await upl->initialize(range, deadline);
+    if (res.has_failure()) {
+        co_return res.as_failure();
+    }
+    co_return std::move(upl);
+}
+
+ss::future<result<std::unique_ptr<segment_upload>>>
+segment_upload::make_segment_upload(
+  ss::lw_shared_ptr<cluster::partition> part,
+  size_limited_offset_range range,
+  size_t read_buffer_size,
+  ss::scheduling_group sg,
+  model::timeout_clock::time_point deadline) {
+    std::unique_ptr<segment_upload> upl(
+      new segment_upload(part, read_buffer_size, sg));
+
+    auto res = co_await upl->initialize(range, deadline);
+    if (res.has_failure()) {
+        co_return res.as_failure();
+    }
+    co_return std::move(upl);
+}
+
+ss::future<result<void>> segment_upload::initialize(
+  inclusive_offset_range range, model::timeout_clock::time_point deadline) {
+    auto holder = _gate.hold();
+    auto params = co_await compute_upload_parameters(range, deadline);
+    if (params.has_failure()) {
+        co_return params.as_failure();
+    }
+    _params = std::move(params.value());
+    // Create a log reader config to scan the uploaded offset
+    // range. We should skip the batch cache.
+    storage::log_reader_config reader_cfg(
+      range.base, range.last, priority_manager::local().archival_priority());
```

Comment:
In `compute_upload_parameters` which takes the range as input, it returns a new offset range as part of `upload_reconciliation_result` which looks like it gets stored in `_params`. Would it make sense to use the `last` field from `_params` here? Or are these two always guaranteed to be the same?